File: 42-trapping_rain_water.py
def trap(height):
    """
    Computes the total amount of trapped rainwater given elevation heights.
    :param height: List[int] - List of non-negative integers representing elevation heights.
    :return: int - Total units of trapped rainwater.
    """
    if not height or len(height) < 3:
        return 0  # If there are less than 3 bars, no water can be trapped.
    
    n = len(height)
    left_max = [0] * n  # Array to store maximum height to the left of each index.
    right_max = [0] * n  # Array to store maximum height to the right of each index.
    
    # Fill left_max array
    left_max[0] = height[0]
    for i in range(1, n):
        left_max[i] = max(left_max[i - 1], height[i])
    
    # Fill right_max array
    right_max[n - 1] = height[n - 1]
    for i in range(n - 2, -1, -1):
        right_max[i] = max(right_max[i + 1], height[i])
    
    # Calculate trapped water
    total_water = 0
    for i in range(n):
        total_water += min(left_max[i], right_max[i]) - height[i]
    
    return total_water

File: 42-trapping_rain_water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    if (!height || height.length === 0) return 0; // Edge case: empty or null array

    let left = 0, right = height.length - 1; // Pointers to traverse the height array
    let leftMax = 0, rightMax = 0; // Track the maximum height from left and right
    let trappedWater = 0; // Store total trapped water

    while (left < right) {
        if (height[left] < height[right]) {
            // Process the left side
            if (height[left] >= leftMax) {
                leftMax = height[left]; // Update left max
            } else {
                trappedWater += leftMax - height[left]; // Water trapped at this index
            }
            left++; // Move left pointer
        } else {
            // Process the right side
            if (height[right] >= rightMax) {
                rightMax = height[right]; // Update right max
            } else {
                trappedWater += rightMax - height[right]; // Water trapped at this index
            }
            right--; // Move right pointer
        }
    }

    return trappedWater;
};

// Example usage:
console.log(trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1])); // Output: 6
